e2e/stack_monitoring_test.go

```diff
@@ -0,0 +1,614 @@
+		tearDownRuntimeDependencies("stack-monitoring")
+	})
+	s.AfterScenario(func(*messages.Pickle, error) {
+		log.Debug("After StackMonitoring Scenario...")
+		testSuite.removeProduct()
+	})
+}
+
+// assertHitsEqualStructure returns an error if hits don't share structure
+func assertHitsEqualStructure(sm *StackMonitoringTestSuite, collectionMethod1 string, collectionMethod2 string) error {
+	collectionMethod1Hits := sm.collectionHits[collectionMethod1]
+	collectionMethod2Hits := sm.collectionHits[collectionMethod2]
+
+	collectionMethod1HitsJSON := gabs.Wrap(collectionMethod1Hits)
+	collectionMethod2HitsJSON := gabs.Wrap(collectionMethod2Hits)
+
+	err := checkParity(sm, collectionMethod1HitsJSON, collectionMethod2HitsJSON)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func checkParity(sm *StackMonitoringTestSuite, legacyContainer *gabs.Container, metricbeatContainer *gabs.Container) error {
+	allowedInsertionsInMetricbeatDocs := []string{
+		"service",
+		"@timestamp",
+		"agent",
+		"event",
+		"host",
+		"ecs",
+		"metricset",
+	}
+	allowedInsertionsInMetricbeatDocs = append(allowedInsertionsInMetricbeatDocs, sm.allowedInsertionsExtra...)
+
+	allowedDeletionsFromMetricbeatDocs := []string{
+		"source_node",
+	}
+	allowedDeletionsFromMetricbeatDocs = append(allowedDeletionsFromMetricbeatDocs, sm.allowedDeletionsExtra...)
+
+	hitsPath := "hits.hits"
+	legacyHits := legacyContainer.Path(hitsPath)
+	metricbeatHits := metricbeatContainer.Path(hitsPath)
+
+	legacyTypes, legacySources := checkSourceTypes(legacyHits)
+	metricbeatTypes, metricbeatSources := checkSourceTypes(metricbeatHits)
+
+	if len(legacyTypes) > len(metricbeatTypes) {
+		// returns an array as the result of removing all elements in 'b' from the 'a' array
+		// only used here
+		arrayDiff := func(a []string, b []string) []string {
+			target := map[string]bool{}
+			for _, x := range b {
+				target[x] = true
+			}
+
+			result := []string{}
+			for _, x := range a {
+				if _, ok := target[x]; !ok {
+					result = append(result, x)
+				}
+			}
+
+			return result
+		}
+
+		diff := arrayDiff(legacyTypes, metricbeatTypes)
+
+		return fmt.Errorf("Found more legacy-indexed document types than metricbeat-indexed document types. Document types indexed by internal collection but not by Metricbeat collection: %v", diff)
+	}
+
+	errors := []error{}
+	for docType, sourceValue := range legacySources {
+		legacyDoc := gabs.Wrap(sourceValue)
+		metricbeatDoc := gabs.Wrap(metricbeatSources[docType])
+
+		err := sm.handleSpecialCases(docType, legacyDoc, metricbeatDoc)
+		if err != nil {
+			errors = append(errors, err)
+		}
+
+		unexpectedInsertions := []string{}
+		unexpectedDeletions := []string{}
+
+		differences := jsonDiff(legacyDoc, metricbeatDoc)
+
+		lines := strings.Split(differences, "\n")
+		for _, line := range lines {
+			trimmedLine := strings.TrimSpace(line)
+
+			if !strings.HasPrefix(trimmedLine, `"ADDED":`) && !strings.HasPrefix(trimmedLine, `"REMOVED":`) {
+				continue
```

Comment:
For each of these `continue` cases, it might be helpful to have a comment _why_ we're OK to continue in each case.

Reply:
Added in latest commit